### PR TITLE
fix linkify

### LINKED ISSUE #2038: Messages are not linkified anymore
- send a message containing a link
- observe that the link is not rendered correctly and not clickable

image

### PR DESCRIPTION
**Pull request by bmarty**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Ensure message with body and without formatted body got working links.
Also ensure email addresses are linkified.

## Motivation and context

Closes #2038

## Tests

- Send a message with just a raw link and nothing else
- Observe that with this change the link in the message is rendered as a link and is clickable

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt [hunk 1]
```diff
@@ -68,12 +69,15 @@ class TimelineItemContentMessageFactory @Inject constructor(
...
-                htmlDocument = messageType.formatted?.toHtmlDocument(prefix = "* $senderDisplayName"),
-                formattedBody = parseHtml(messageType.formatted, prefix = "* $senderDisplayName"),
-                isEdited = content.isEdited,
-            )
+            is EmoteMessageType -> {
+                val emoteBody = "* $senderDisplayName ${messageType.body}"
+                TimelineItemEmoteContent(
+                    body = emoteBody,
+                    htmlDocument = messageType.formatted?.toHtmlDocument(prefix = "* $senderDisplayName"),
+                    formattedBody = parseHtml(messageType.formatted, prefix = "* $senderDisplayName") ?: emoteBody.withLinks(),
```
Comment (ganfra) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt:77]: If I understand correctly, we always have a `formattedBody` now with those fixes. Could we simplify a bit the `TimelineItemTextBasedContent` interface then? 
It's a bit hard to know what to use currently, and we probably don't need all the fields.
```
sealed interface TimelineItemTextBasedContent : TimelineItemEventContent {
    val body: String
    val htmlDocument: Document?
    val formattedBody: CharSequence?
    val plainText: String
    val isEdited: Boolean
    val htmlBody: String?
        get() = htmlDocument?.body()?.html()
}
```

**Comment by frebib** [comment]
Would we ever consider also detecting and formatting raw user/room mentions in this way too, for clients that only send unformatted bodies?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt (hunk 1 above)
Comment (bmarty) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt:77]: > we always have a `formattedBody`

No, `withLinks()` returns `null` if no links are added.

File: appnav/src/test/kotlin/io/element/android/appnav/loggedin/LoggedInPresenterTest.kt [hunk 2]
```diff
@@ -60,8 +60,9 @@ class LoggedInPresenterTest {
         }.test {
             val initialState = awaitItem()
             assertThat(initialState.showSyncSpinner).isFalse()
+            roomListService.postSyncIndicator(RoomListService.SyncIndicator.Show)
             consumeItemsUntilPredicate { it.showSyncSpinner }
-            roomListService.postState(RoomListService.State.Running)
+            roomListService.postSyncIndicator(RoomListService.SyncIndicator.Hide)
```
Comment (bmarty) [inline appnav/src/test/kotlin/io/element/android/appnav/loggedin/LoggedInPresenterTest.kt:65]: `consumeItemsUntilPredicate` now expect the predicate to be true at some point, so I had to fix this outdated test.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt (hunk 1 above)
Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt:77]: Can we still try to simplify the api? 
If we could get a `plainText/formattedText/body/htmlBody` and remove the `htmlDocument` instance?
I'd say, The `formattedText` shouldn't be nullable and should always be displayed in timeline.

Comment (bmarty) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt:77]: Yes, but I would prefer to do that in a separate rework PR. OK for you?

Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/factories/event/TimelineItemContentMessageFactory.kt:77]: as u want

**Review by ganfra — APPROVED** [review]

**Pull request merged**